File: Shared/Models/HikeStatistics.swift
// Copyright (C) 2024-2026 Dr Horst Herb
//
// This file is part of OpenHiker.
//
// OpenHiker is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// OpenHiker is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with OpenHiker. If not, see <https://www.gnu.org/licenses/>.

import Foundation

/// A value type aggregating all statistics for a single hike.
///
/// This is a shared infrastructure type used by multiple subsystems:
/// - **Stats overlay** on the watch map (``HikeStatsOverlay``)
/// - **HealthKit** workout recording (``HealthKitManager``)
/// - **Saved routes** (Phase 3) for persistent hike records
/// - **Export** (Phase 6) for PDF/Markdown hike reports
///
/// All distance values are in meters, all time values are in seconds,
/// and all speed values are in meters per second. The UI layer is
/// responsible for locale-aware formatting (metric vs. imperial).
struct HikeStatistics: Codable, Sendable, Equatable {

    // MARK: - Distance & Elevation

    /// Total distance covered in meters, measured along the GPS track.
    let totalDistance: Double

    /// Total cumulative elevation gained in meters (only uphill segments).
    let elevationGain: Double

    /// Total cumulative elevation lost in meters (only downhill segments).
    let elevationLoss: Double

    // MARK: - Time

    /// Total elapsed time from hike start to finish, in seconds.
    let duration: TimeInterval

    /// Time spent actively moving (speed >= ``HikeStatisticsConfig/restingSpeedThreshold``), in seconds.
    let walkingTime: TimeInterval

    /// Time spent stationary or nearly stationary, in seconds.
    let restingTime: TimeInterval

    // MARK: - Heart Rate & SpO2 (optional, requires HealthKit)

    /// Average heart rate during the hike in beats per minute, or `nil` if HealthKit data is unavailable.
    let averageHeartRate: Double?

    /// Maximum heart rate recorded during the hike in beats per minute, or `nil` if unavailable.
    let maxHeartRate: Double?

    /// Average blood oxygen saturation percentage during the hike, or `nil` if unavailable.
    let averageSpO2: Double?

    // MARK: - Energy

    /// Estimated energy burned in kilocalories.
    ///
    /// Always has a value because ``CalorieEstimator`` falls back to
    /// ``HikeStatisticsConfig/defaultBodyMassKg`` when HealthKit body mass is unavailable.
    let estimatedCalories: Double?

    // MARK: - Speed

    /// Average speed in meters per second over the entire hike duration.
    let averageSpeed: Double

    /// Maximum instantaneous speed recorded in meters per second.
    let maxSpeed: Double
}

// MARK: - Configuration

/// Configuration constants for hike statistics calculations.
///
/// These constants control how statistics are computed from raw GPS and
/// HealthKit data. They are defined as static constants rather than
/// magic numbers to make the code self-documenting and easy to adjust.
enum HikeStatisticsConfig {

    /// Speed threshold in m/s below which the hiker is considered resting.
    ///
    /// 0.3 m/s is roughly 1 km/h — below normal walking pace, accounting
    /// for GPS drift when standing still.
    static let restingSpeedThreshold: Double = 0.3

    /// Base MET (Metabolic Equivalent of Task) value for moderate hiking.
    ///
    /// Source: Compendium of Physical Activities — "hiking, cross country" = 6.0 MET.
    static let baseHikingMET: Double = 6.0

    /// Additional MET per 10% average grade to account for steeper terrain.
    ///
    /// For every 10% of average grade (elevation gain / horizontal distance),
    /// add 1.0 MET to the base value.
    static let metPerTenPercentGrade: Double = 1.0

    /// Default body mass in kilograms, used when HealthKit body mass is unavailable.
    ///
    /// 70 kg is the WHO reference adult weight.
    static let defaultBodyMassKg: Double = 70.0

    /// Maximum age in seconds for a SpO2 reading to be considered "recent".
    ///
    /// Blood oxygen samples older than this threshold are hidden from the
    /// live stats overlay since they may no longer reflect current status.
    static let spO2MaxAgeSec: TimeInterval = 300.0

    // MARK: - Unit Conversion Constants

    /// Meters per mile (international mile).
    static let metersPerMile: Double = 1609.344

    /// Feet per meter (international foot).
    static let feetPerMeter: Double = 3.28084

    /// Conversion factor from m/s to km/h.
    static let kmhPerMps: Double = 3.6

    /// Conversion factor from m/s to mph.
    static let mphPerMps: Double = 2.23694
}

// MARK: - Formatting Helpers

/// Provides locale-aware formatting for hike statistics values.
///
/// All formatting respects the user's metric/imperial preference stored in
/// `@AppStorage("useMetricUnits")`. The formatters are designed for compact
/// display on Apple Watch screens.
enum HikeStatsFormatter {

    /// Formats a distance in meters as a human-readable string.
    ///
    /// - Parameters:
    ///   - meters: The distance in meters.
    ///   - useMetric: If `true`, formats as km; if `false`, formats as mi.
    /// - Returns: A formatted string like "3.2 km" or "2.0 mi".
    static func formatDistance(_ meters: Double, useMetric: Bool) -> String {
        if useMetric {
            let km = meters / 1000.0
            return String(format: "%.1f km", km)
        } else {
            let miles = meters / HikeStatisticsConfig.metersPerMile
            return String(format: "%.1f mi", miles)
        }
    }

    /// Formats an elevation value in meters as a human-readable string.
    ///
    /// - Parameters:
    ///   - meters: The elevation in meters.
    ///   - useMetric: If `true`, formats as m; if `false`, formats as ft.
    /// - Returns: A formatted string like "450 m" or "1476 ft".
    static func formatElevation(_ meters: Double, useMetric: Bool) -> String {
        if useMetric {
            return String(format: "%.0f m", meters)
        } else {
            let feet = meters * HikeStatisticsConfig.feetPerMeter
            return String(format: "%.0f ft", feet)
        }
    }

    /// Formats a time interval as HH:MM:SS.
    ///
    /// - Parameter interval: The time interval in seconds.
    /// - Returns: A formatted string like "02:15:30" or "00:45:12".
    static func formatDuration(_ interval: TimeInterval) -> String {
        let totalSeconds = Int(interval)
        let hours = totalSeconds / 3600
        let minutes = (totalSeconds % 3600) / 60
        let seconds = totalSeconds % 60
        return String(format: "%02d:%02d:%02d", hours, minutes, seconds)
    }

    /// Formats a speed in m/s as a human-readable string.
    ///
    /// - Parameters:
    ///   - metersPerSecond: The speed in meters per second.
    ///   - useMetric: If `true`, formats as km/h; if `false`, formats as mph.
    /// - Returns: A formatted string like "4.5 km/h" or "2.8 mph".
    static func formatSpeed(_ metersPerSecond: Double, useMetric: Bool) -> String {
        if useMetric {
            let kmh = metersPerSecond * HikeStatisticsConfig.kmhPerMps
            return String(format: "%.1f km/h", kmh)
        } else {
            let mph = metersPerSecond * HikeStatisticsConfig.mphPerMps
            return String(format: "%.1f mph", mph)
        }
    }

    /// Formats a heart rate value as a BPM string.
    ///
    /// - Parameter bpm: Heart rate in beats per minute.
    /// - Returns: A formatted string like "142 bpm".
    static func formatHeartRate(_ bpm: Double) -> String {
        return String(format: "%.0f bpm", bpm)
    }

    /// Formats a blood oxygen saturation value as a percentage string.
    ///
    /// - Parameter fraction: SpO2 as a fraction (0.0 to 1.0).
    /// - Returns: A formatted string like "97%".
    static func formatSpO2(_ fraction: Double) -> String {
        return String(format: "%.0f%%", fraction * 100)
    }

    /// Formats a calorie count as a human-readable string.
    ///
    /// - Parameter kcal: Energy in kilocalories.
    /// - Returns: A formatted string like "523 kcal".
    static func formatCalories(_ kcal: Double) -> String {
        return String(format: "%.0f kcal", kcal)
    }
}

// MARK: - Calorie Estimation

/// Pure-function calorie estimator using MET-based calculation.
///
/// This uses the standard MET formula from exercise physiology:
/// `calories = MET * bodyMassKg * durationHours`
///
/// The base MET for hiking (6.0) is adjusted upward for steeper terrain
/// using the average grade derived from elevation gain and horizontal distance.
enum CalorieEstimator {

    /// Estimates calories burned during a hike.
    ///
    /// - Parameters:
    ///   - distanceMeters: Total horizontal distance in meters.
    ///   - elevationGainMeters: Total cumulative elevation gain in meters.
    ///   - durationSeconds: Total hike duration in seconds.
    ///   - bodyMassKg: Hiker's body mass in kilograms. Uses ``HikeStatisticsConfig/defaultBodyMassKg``
    ///     (70 kg) if `nil`.
    /// - Returns: Estimated energy burned in kilocalories.
    static func estimateCalories(
        distanceMeters: Double,
        elevationGainMeters: Double,
        durationSeconds: TimeInterval,
        bodyMassKg: Double?
    ) -> Double {
        let mass = bodyMassKg ?? HikeStatisticsConfig.defaultBodyMassKg
        let durationHours = durationSeconds / 3600.0

        // Calculate average grade as a percentage
        var met = HikeStatisticsConfig.baseHikingMET
        if distanceMeters > 0 {
            let gradePercent = (elevationGainMeters / distanceMeters) * 100.0
            let gradeTens = gradePercent / 10.0
            met += gradeTens * HikeStatisticsConfig.metPerTenPercentGrade
        }

        return met * mass * durationHours
    }
}

File: OpenHikerAndroid/core/src/main/kotlin/com/openhiker/core/model/HikeStatistics.kt
/*
 * OpenHiker - Offline Hiking Navigation
 * Copyright (C) 2024 - 2026 Dr Horst Herb
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.openhiker.core.model

import kotlinx.serialization.Serializable

/**
 * Accumulated statistics for a completed hike.
 *
 * All distance values are in metres, all times in seconds, speeds in m/s.
 * Optional fields (heart rate, SpO2, calories) are null when the data
 * source is unavailable (e.g., no wearable connected).
 *
 * @property totalDistance Total distance walked in metres (Haversine sum).
 * @property elevationGain Cumulative uphill elevation change in metres.
 * @property elevationLoss Cumulative downhill elevation change in metres.
 * @property duration Total elapsed time in seconds (start to finish).
 * @property walkingTime Time spent moving (speed > resting threshold) in seconds.
 * @property restingTime Time spent stationary in seconds.
 * @property averageHeartRate Mean heart rate in BPM, or null if unavailable.
 * @property maxHeartRate Peak heart rate in BPM, or null if unavailable.
 * @property averageSpO2 Mean blood oxygen as fraction (0.0–1.0), or null.
 * @property estimatedCalories Estimated kilocalories burned, or null.
 * @property averageSpeed Mean moving speed in m/s.
 * @property maxSpeed Peak speed in m/s.
 */
@Serializable
data class HikeStatistics(
    val totalDistance: Double = 0.0,
    val elevationGain: Double = 0.0,
    val elevationLoss: Double = 0.0,
    val duration: Double = 0.0,
    val walkingTime: Double = 0.0,
    val restingTime: Double = 0.0,
    val averageHeartRate: Double? = null,
    val maxHeartRate: Double? = null,
    val averageSpO2: Double? = null,
    val estimatedCalories: Double? = null,
    val averageSpeed: Double = 0.0,
    val maxSpeed: Double = 0.0
)

/**
 * Configuration constants for hike statistics calculations.
 *
 * All thresholds and conversion factors are defined here to avoid
 * magic numbers throughout the codebase. Values match the iOS
 * HikeStatisticsConfig for consistent behaviour across platforms.
 */
object HikeStatisticsConfig {
    /** Speed below which the user is considered resting, in m/s (~1 km/h). */
    const val RESTING_SPEED_THRESHOLD = 0.3

    /** Base MET value for hiking (Compendium of Physical Activities). */
    const val BASE_HIKING_MET = 6.0

    /** Additional MET per 10% grade increase. */
    const val MET_PER_TEN_PERCENT_GRADE = 1.0

    /** Default body mass for calorie estimation (WHO reference), in kg. */
    const val DEFAULT_BODY_MASS_KG = 70.0

    /** Minimum rest duration to count as a rest period, in seconds. */
    const val MIN_REST_DURATION_SEC = 60.0

    /** Maximum age of SpO2 reading before it's considered stale, in seconds. */
    const val SPO2_MAX_AGE_SEC = 300.0

    /** Metres per statute mile. */
    const val METRES_PER_MILE = 1609.344

    /** Feet per metre. */
    const val FEET_PER_METRE = 3.28084

    /** km/h per m/s. */
    const val KMH_PER_MPS = 3.6

    /** mph per m/s. */
    const val MPH_PER_MPS = 2.23694

    /** Easy effort heart rate ceiling in BPM. */
    const val EASY_HEART_RATE_MAX = 120.0

    /** Moderate effort heart rate ceiling in BPM. */
    const val MODERATE_HEART_RATE_MAX = 150.0

    /** Metres per kilometre. */
    const val METRES_PER_KILOMETRE = 1000.0
}

/**
 * Pure functions for formatting hike statistics for display.
 *
 * All formatters accept raw metric values and return formatted strings.
 * The [useMetric] parameter controls metric vs. imperial output.
 */
object HikeStatsFormatter {

    /**
     * Formats a distance value for display.
     *
     * @param metres Distance in metres.
     * @param useMetric True for km, false for miles.
     * @return Formatted string like "3.2 km" or "2.0 mi".
     */
    fun formatDistance(metres: Double, useMetric: Boolean = true): String {
        return if (useMetric) {
            "%.1f km".format(metres / HikeStatisticsConfig.METRES_PER_KILOMETRE)
        } else {
            "%.1f mi".format(metres / HikeStatisticsConfig.METRES_PER_MILE)
        }
    }

    /**
     * Formats an elevation value for display.
     *
     * @param metres Elevation in metres.
     * @param useMetric True for metres, false for feet.
     * @return Formatted string like "450 m" or "1476 ft".
     */
    fun formatElevation(metres: Double, useMetric: Boolean = true): String {
        return if (useMetric) {
            "%.0f m".format(metres)
        } else {
            "%.0f ft".format(metres * HikeStatisticsConfig.FEET_PER_METRE)
        }
    }

    /**
     * Formats a duration as HH:MM:SS.
     *
     * @param seconds Duration in seconds.
     * @return Formatted string like "02:34:15".
     */
    fun formatDuration(seconds: Double): String {
        val totalSeconds = seconds.toLong()
        val hours = totalSeconds / 3600
        val minutes = (totalSeconds % 3600) / 60
        val secs = totalSeconds % 60
        return "%02d:%02d:%02d".format(hours, minutes, secs)
    }

    /**
     * Formats a speed value for display.
     *
     * @param metersPerSecond Speed in m/s.
     * @param useMetric True for km/h, false for mph.
     * @return Formatted string like "4.5 km/h" or "2.8 mph".
     */
    fun formatSpeed(metersPerSecond: Double, useMetric: Boolean = true): String {
        return if (useMetric) {
            "%.1f km/h".format(metersPerSecond * HikeStatisticsConfig.KMH_PER_MPS)
        } else {
            "%.1f mph".format(metersPerSecond * HikeStatisticsConfig.MPH_PER_MPS)
        }
    }

    /**
     * Formats a heart rate value.
     *
     * @param bpm Heart rate in beats per minute.
     * @return Formatted string like "142 bpm".
     */
    fun formatHeartRate(bpm: Double): String = "%.0f bpm".format(bpm)

    /**
     * Formats a blood oxygen saturation value.
     *
     * @param fraction SpO2 as a fraction (0.0–1.0).
     * @return Formatted string like "97%".
     */
    fun formatSpO2(fraction: Double): String = "%.0f%%".format(fraction * 100.0)

    /**
     * Formats a calorie count.
     *
     * @param kcal Kilocalories burned.
     * @return Formatted string like "523 kcal".
     */
    fun formatCalories(kcal: Double): String = "%.0f kcal".format(kcal)
}

/**
 * Pure function for estimating calories burned during a hike.
 *
 * Uses the MET (Metabolic Equivalent of Task) method with grade adjustment.
 * Formula: MET * bodyMass * durationHours, where MET increases with steeper grades.
 */
object CalorieEstimator {

    /**
     * Estimates calories burned during a hike.
     *
     * @param distanceMetres Total distance walked in metres.
     * @param elevationGainMetres Total elevation gained in metres.
     * @param durationSeconds Total duration in seconds.
     * @param bodyMassKg Body mass in kilograms (defaults to 70 kg WHO reference).
     * @return Estimated kilocalories burned.
     */
    fun estimate(
        distanceMetres: Double,
        elevationGainMetres: Double,
        durationSeconds: Double,
        bodyMassKg: Double = HikeStatisticsConfig.DEFAULT_BODY_MASS_KG
    ): Double {
        if (durationSeconds <= 0.0 || distanceMetres <= 0.0) return 0.0

        val gradePercent = if (distanceMetres > 0) {
            (elevationGainMetres / distanceMetres) * 100.0
        } else {
            0.0
        }

        val met = HikeStatisticsConfig.BASE_HIKING_MET +
            (gradePercent / 10.0) * HikeStatisticsConfig.MET_PER_TEN_PERCENT_GRADE
        val durationHours = durationSeconds / 3600.0

        return met * bodyMassKg * durationHours
    }
}
